pkg/app/piped/livestatereporter/cloudrun/report.go

```diff
@@ -0,0 +1,134 @@
+}
+
+type apiClient interface {
+	ReportApplicationLiveState(ctx context.Context, req *pipedservice.ReportApplicationLiveStateRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateResponse, error)
+	ReportApplicationLiveStateEvents(ctx context.Context, req *pipedservice.ReportApplicationLiveStateEventsRequest, opts ...grpc.CallOption) (*pipedservice.ReportApplicationLiveStateEventsResponse, error)
+}
+
+type Reporter interface {
+	Run(ctx context.Context) error
+	ProviderName() string
+}
+
+type reporter struct {
+	provider              config.PipedCloudProvider
+	appLister             applicationLister
+	stateGetter           cloudrun.Getter
+	apiClient             apiClient
+	snapshotFlushInterval time.Duration
+	logger                *zap.Logger
+
+	snapshotVersions map[string]model.ApplicationLiveStateVersion
+}
+
+func NewReporter(cp config.PipedCloudProvider, appLister applicationLister, stateGetter cloudrun.Getter, apiClient apiClient, logger *zap.Logger) Reporter {
+	logger = logger.Named("cloudrun-reporter").With(
+		zap.String("cloud-provider", cp.Name),
+	)
+	return &reporter{
+		provider:              cp,
+		appLister:             appLister,
+		stateGetter:           stateGetter,
+		apiClient:             apiClient,
+		snapshotFlushInterval: time.Minute,
+		logger:                logger,
+		snapshotVersions:      make(map[string]model.ApplicationLiveStateVersion),
+	}
+}
+
+func (r *reporter) Run(ctx context.Context) error {
+	r.logger.Info("start running app live state reporter")
+
+	r.logger.Info("waiting for livestatestore to be ready")
+	if err := r.stateGetter.WaitForReady(ctx, 10*time.Minute); err != nil {
+		r.logger.Error("livestatestore was unable to be ready in time", zap.Error(err))
+		return err
+	}
+
+	snapshotTicker := time.NewTicker(r.snapshotFlushInterval)
+	defer snapshotTicker.Stop()
+
+	for {
+		select {
+		case <-snapshotTicker.C:
+			r.flushSnapshots(ctx)
+
+		case <-ctx.Done():
+			r.logger.Info("app live state reporter has been stopped")
+			return nil
+		}
+	}
+}
+
+func (r *reporter) ProviderName() string {
+	return r.provider.Name
+}
+
+func (r *reporter) flushSnapshots(ctx context.Context) error {
+	apps := r.appLister.ListByCloudProvider(r.provider.Name)
+	for _, app := range apps {
+		state, ok := r.stateGetter.GetState(app.Id)
+		if !ok {
+			r.logger.Info(fmt.Sprintf("no app state of cloudrun application %s to report", app.Id))
+			continue
+		}
+
+		snapshot := &model.ApplicationLiveStateSnapshot{
+			ApplicationId: app.Id,
+			PipedId:       app.PipedId,
+			ProjectId:     app.ProjectId,
+			Kind:          app.Kind,
+			Cloudrun: &model.CloudRunApplicationLiveState{
+				Resources: state.Resources,
+			},
+			Version: &state.Version,
+		}
+		snapshot.DetermineAppHealthStatus()
+		req := &pipedservice.ReportApplicationLiveStateRequest{
+			Snapshot: snapshot,
+		}
+
+		if _, err := r.apiClient.ReportApplicationLiveState(ctx, req); err != nil {
```

Comment:
nit: We may need a new RPC to send a bunch of applications.

Reply:
Yep, It may be good to add new RPC to improve performance in the future.